### PR TITLE
FuncDeclaration: Move booleans to FUNCFLAG

### PR DESCRIPTION
**Pull request by Geod24**
```
FUNCFLAG was intended to store boolean, but a few cropped back up.
Doing this reduces the size of FuncDeclaration from 656 to 624 bytes.
```

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by WalterBright** [comment]
A nice demonstration of why D should have bit fields.

**Comment by Geod24** [comment]
I cannot agree with you here.

File: src/dmd/func.d [hunk 1]
```diff
@@ -205,6 +205,17 @@ enum FUNCFLAG : uint
...
+    naked            = 0x2000, /// The function is 'naked' (see inline ASM)
+    generated        = 0x4000, /// The function is compiler generated (e.g. `opCmp`)
+    introducing      = 0x8000, /// If this function introduces the overload set
+    semantic3Errors  = 0x10000, /// If errors in semantic3 this function's frame ptr
+    noEH             = 0x20000, /// No exception unwinding is needed
+    inferRetType     = 0x40000, /// Return type is to be inferred
+    dualContext      = 0x80000, /// has a dual-context 'this' parameter
+    hasAlwaysInline  = 0x100000, /// Contains references to functions that must be inlined
+    CRTCtor          = 0x200000, /// Has attribute pragma(crt_constructor) (set in glue)
+    CRTDtor          = 0x400000, /// Has attribpute pragma(crt_destructor) (set in glue)
```
Comment (ibuclaw) [inline src/dmd/func.d:218]: Missing from headers.

File: src/dmd/func.d [hunk 2]
```diff
@@ -1482,6 +1485,51 @@ extern (C++) class FuncDeclaration : Declaration
...
+
+    final bool hasDualContext() const scope @safe pure nothrow @nogc
+    {
+        return !!(this.flags & FUNCFLAG.dualContext);
+    }
+
+    final bool hasAlwaysInlines() const scope @safe pure nothrow @nogc
+    {
+        return !!(this.flags & FUNCFLAG.hasAlwaysInline);
+    }
```
Comment (ibuclaw) [inline src/dmd/func.d:1531]: Missing from headers.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

**Comment by ibuclaw** [comment]
Thanks for breaking downstream with the lightning merge.

File: src/dmd/func.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/func.d:218]: This is not part of the C++ headers though (FUNCFLAG itself)

**Comment by Geod24** [comment]
https://github.com/dlang/dmd/pull/13858

File: src/dmd/func.d (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/func.d:218]: Ok, so just the accessors are enough then to allow gdc/ldc to continue querying these.

File: src/dmd/glue.d [hunk 3]
```diff
@@ -867,17 +868,17 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
         fd.shidden = shidden;
     }
     else
     {
         // Register return style cannot make nrvo.
-        // Auto functions keep the nrvo_can flag up to here,
+        // Auto functions keep the NRVO flag up to here,
         // so we should eliminate it before entering backend.
-        fd.nrvo_can = 0;
+        fd.flags &= ~FUNCFLAG.NRVO;
```
Comment (ibuclaw) [inline src/dmd/glue.d:881]: Still broken (this is a setter that might need exposing to C++).

File: src/dmd/frontend.h [hunk 4]
```diff
@@ -2883,6 +2873,15 @@ class FuncDeclaration : public Declaration
     bool isTrusted();
     bool isNogc();
     bool isNogcBypassingInference();
+    bool isNRVO() const;
+    bool isNaked() const;
+    bool isGenerated() const;
```
Comment (ibuclaw) [inline src/dmd/frontend.h:2878]: Also missing a setter for this field.

**Pull request merged**